clippy_lints/src/operators/manual_midpoint.rs

```diff
@@ -0,0 +1,45 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::msrvs::{self, Msrv};
+use clippy_utils::sugg::Sugg;
+use clippy_utils::{is_floating_point_integer_literal, is_integer_literal};
+use rustc_ast::BinOpKind;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::LateContext;
+use rustc_middle::ty;
+
+use super::MANUAL_MIDPOINT;
+
+pub(super) fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx Expr<'_>,
+    op: BinOpKind,
+    left: &'tcx Expr<'_>,
+    right: &'tcx Expr<'_>,
+    msrv: &Msrv,
+) {
+    if msrv.meets(msrvs::UINT_FLOAT_MIDPOINT)
+        && op == BinOpKind::Div
+        && (is_integer_literal(right, 2) || is_floating_point_integer_literal(right, 2))
+        && let ExprKind::Binary(left_op, ll_expr, lr_expr) = left.kind
+        && left_op.node == BinOpKind::Add
+        && let left_ty = cx.typeck_results().expr_ty_adjusted(ll_expr)
+        && let right_ty = cx.typeck_results().expr_ty_adjusted(lr_expr)
+        && left_ty == right_ty
+        && matches!(left_ty.kind(), ty::Uint(_) | ty::Float(_))
+    {
+        let mut app = Applicability::MachineApplicable;
+        let left_sugg = Sugg::hir_with_applicability(cx, ll_expr, "..", &mut app);
+        let right_sugg = Sugg::hir_with_applicability(cx, lr_expr, "..", &mut app);
+        let sugg = format!("{left_ty}::midpoint({left_sugg}, {right_sugg})");
```

Comment:
Interesting choice of form, I would have probably use the method syntax (ie. `.midpoint(..)`) instead, as to not have to much shifts from the original code, but that works as well.

Reply:
I feel that linting `(a + b) / 2` as `a.midpoint(b)` introduces an asymmetry between `a` and `b` while `u32::midpoint(a, b)`  doesn't. But I'm not opposed to this change.

What do others think? Please upvote if you prefer `a.midpoint(b)` and downvote if you prefer `u32::midpoint(a, b)`.

Reply:
I can't pick between the 2 because the first one is more convenirnt but the second one is more readable...

Well, by writing this comment I realized I prefer more readable code so let's go for 2.